_posts/contests/2021-02-05-Performance-analysis-and-tuning-contest-4.md

```diff
@@ -0,0 +1,104 @@
+  * Make sure innermost loop has the high trip count. This will allow compiler optimizations like loop unrolling, vectorization, etc. to speed up the code processing.
+  * Move all loop-invariant code outside of the loop. Make innermost loop as compact as possible. 
+  * Run your code through [TMAM]({{ site.url }}/blog/2019/02/09/ or Intel's VTUNE profiler to understand which line cause the CPU to get stalled (due to cache misses, branch misspredictions, etc.)
+4. Fix the issue, build the benchmark, run it and compare against baseline.
+5. Repeat steps 2-5.
+
+Canny is a typical image processing algorithm that runs through the image, sometimes row-wise, sometimes column-wise, and processes pixels. Processing is done in several stages. Collecting the performance profile will help you focus on the right functions; collecting information about stalled cycles will help you understand why that code is slow.
+
+I also have a few general advises:
+- **Do not try to understand the whole algorithm**. For some people it's crucial to understand how every peace of code works. For the purposes of optimizing it will be wasted effort. There are CPU benchmarks with thousands LOC (like [SPEC2017](http://spec.org/cpu2017/)) it's absolutely impossible to understand them in a reasonable time. What you need to familiarize yourself with are hotspots. That's it. You most likely need to understand one function/loop which is not more than 100 LOC.
+- **You have specific workload for which you optimize the benchmark**. You don't need to optimize it for any other input/workload. The main principle behind [Data-oriented design](https://en.wikipedia.org/wiki/Data-oriented_design) is that you know the data of your application.
+
+If you feel you're stuck, don't hesitate to ask questions or look for support elsewhere. I don't have much time to answer every question promptly, but I will do my best. You can send questions to me directly using the [contact form on my web site](https://johnysswlab.com/contact/) or to Denis (TODO: Add link)
+
+__See the Q&A post about what optimizations are [allowed]({{ site.url }}/blog/2019/02/02/Performance-optimization-contest#q5-what-optimizations-are-allowed) and what [not]({{ site.url }}/blog/2019/02/02/Performance-optimization-contest#q6-whats-not-allowed).__
+
+### Validation
+
+If the produced image is correct it will print `Validation successful`. A slight tolerance between the reference output image and the image produced by your algorithm is allowed in order to fully exploit the hardaware's resources.
+
+### Submissions
+
+> __Disclaimer__: We will not use submissions for any commercial purposes. However, we can use the submissions for educational purposes.
+
+The baseline that I will be measuring against is 'clang -O3 -march=core-avx2').
+
+If you're willing to submit your work __subscribe to Denis' [mailing list]({{ page.url }}#mc_embed_signup)__ and then send all that you have via email.
```

Comment:
Is it okay if I will add your email to the message that I will be sending from my emailing service? Something like:

"If you have any questions, reach out Ivica at <your email>. Submit your solution to the same email."

Reply:
Yes, you can add it. Use the email: ivica@johnysswlab.com